Preview downsample

GrahamByrnes opened the pull request:
Attempt number 4.

Reminder: @bastibe had the good idea to reduce the computational load by implementing downsampling of the preview image: that's the smallish one at the upper-left of the screen in darkroom view. In current dt, it takes about the same time to compute image updates for this as for the main image (typically around 450ms on my W10). By reducing the linear scale by 0.5 or 0.25, the time spent waiting for the preview image is drastically reduced. There is a noticeable reduction in the time spent waiting for the "working" label to go out, which makes image modifications a little more "real time".

Note that the benefit will depend on your machine: if you have OpenCL and a fast GPU, the preview and main images will each be computed separately in parallel, so in principle, no gain. On the other hand, a larger screen is possibly going to slow things down and provide more room for improvement: I don't know, I have 2K-ish screen and only the CPU.

There are also circumstances in which you would not want to use this (so set preview_downsampling=1 in Core Settings): where you are trying to construct masks which closely follow edges in the image. As pointed out by @aurelienpierre , taking maximal or minimal values of averages generates biased estimates. It's not something I do very often. It's easy to change the setting, and to answer a question posed by @Nilvus yes, it changes without shutting down... although you're likely to need to exit darkroom to lighttable and re-enter.

For the technical side, the clever programming was stolen from @bastibe and consists of a relatively small number of lines in develop.c and pixelpipe_hb.c, which look after the transfoms made going up & down the pixelpipes. The rest was bringing brushes.c and circle.c in-line with the other masks, which extract the downsampling factor from the pixelpipe transforms: I borrowed a bunch of lines from ellipse.c and transfused them into circle.c and brush.c.

The only other change is to darktableconfig.xml.in, which provides the option to change the setting from the options page.

TurboGit commented on src/develop/masks/circle.c:
```diff
@@ -495,51 +495,61 @@ static void dt_circle_events_post_expose(cairo_t *cr, float zoom_scale, dt_masks
-
-      float xpos, ypos;
-      if((gui->posx == -1.f && gui->posy == -1.f) || gui->mouse_leaved_center)
+                         
```
Never more than one empty line to separate code.

TurboGit commented on src/develop/masks/brush.c:
```diff
@@ -2156,8 +2162,9 @@ static void dt_brush_events_post_expose(cairo_t *cr, float zoom_scale, dt_masks_
-      float radius, oldradius, opacity, oldopacity, pressure;
+      float masks_border = 0.0, masks_hardness = 0.0, masks_density = 0.0;
+      float opacity = 0.0, oldopacity = 0.0, pressure = 0.0;
+      float oldradius_hardness = 0;                          
```
0.0f

GrahamByrnes replied:
Ok, I think I got them all, plus some old ones for free :-)

TurboGit commented on src/develop/masks/brush.c:
```diff
@@ -2156,8 +2162,9 @@ static void dt_brush_events_post_expose(cairo_t *cr, float zoom_scale, dt_masks_
     {
-      float masks_border, masks_hardness, masks_density;
-      float radius, oldradius, opacity, oldopacity, pressure;
+      float masks_border = 0.0, masks_hardness = 0.0, masks_density = 0.0;
```
All 0.0 should end with f : 0.0f

This will probably fix the CI builder.

GrahamByrnes replied:
And indeed it did. Thanks!

bastibe commented:
Brilliant! On my computer, this works correctly as far as pickers go.

However, mask drawing seems severely broken. At 1.0, I can't draw masks at all. The mask shapes (regardless of circular, brush, or elliptic) do not move when I move the mouse. At 0.25, they move and I can draw, but the shapes are much larger than the masks as they appear on the image.

Once saved, however, they edit correctly.

Also, Darktable seems to require restarting to accept the new downsampling setting. You can check this by going to the darkroom and zooming in one notch with your scroll wheel, then drag the image around. While dragging, you can see the preview. Attached are two screen shots of a 1.0 preview and a 0.25 preview, recorded as explained:

![Screenshot_20200413_141814](https://user-images.githubusercontent.com/649181/79120217-aed34500-7d92-11ea-9247-4121e66633ab.png)
![Screenshot_20200413_141746](https://user-images.githubusercontent.com/649181/79120222-b2ff6280-7d92-11ea-8c9c-29a8a14ea356.png)

GrahamByrnes commented:
This is strange... on mine it all works :-(
I had diagnostic statements writing to stderr, both from where preview_downsampling is set at the top of develop.c, but also from back_transform_plus: If I changed settings just after start-up, I'd have a string of as-started values for each visible image. Then it swapped to the new value. I could swap it back as well.
But yes, I can verify the difference in pixelation while navigating the screen...

You have a 4K screen... and OpenCL?

GrahamByrnes commented:
So I get can get 0.25 type pixelation while the diagnostics from develop.c are showing prev-down=1.0, and vice-versa.. very strange.

You just made a fresh build from this PR?

bastibe commented:
> You have a 4K screen... and OpenCL?

Correct.

> You just made a fresh build from this PR?

Yes, I checked out your code, deleted my build directory, and rebuilt from scratch. Do I need to clear anything else beyond the build directory?

GrahamByrnes commented:
No, clearing the build directory should do it. I doubt it could be the screen resolution, maybe the OpenCL makes the difference? 

For the pixelation difference, I don't know what is going on exactly... I'm presuming that there is different information getting to the demosaicing step, which only gets changed after a re-start. I spent a bit of time last week looking for a way of generating a "right now" change over and didn't find a way. I'd be happy to merge and look for a way of improving the switch-over as a stage 2.

However if the masks are not working for you, that's a killer. Obviously I tested all of that. It's particularly strange that you see problems with all masks at 1.0: except for circle and brush, the only difference in the code is multiplying and dividing by 1.0!

GrahamByrnes commented:
Just to be absolutely crystal clear, you're building entirely my code (ie 3.1.0 + (something more than 1133) and not merging into an old master?

bastibe commented:
> Just to be absolutely crystal clear, you're building entirely my code (ie 3.1.0 + (something more than 1133) and not merging into an old master?

Yes, I checked out your code as an entirely new branch, without merging anything into it. As far as I can tell, it's your code and your code alone.

But just to make sure, I'll nuke my entire project directory and download a new copy. Something strange is obviously going on here, and I don't want to hold things up with incompetence ;-).

GrahamByrnes commented:
Now I'm wondering if there was some code I didn't copy across :-(

I'll do a complete rebuild from github as well...

Le mar. 14 avr. 2020 à 09:55, Bastian Bechtold <notifications@github.com> a
écrit :

> Just to be absolutely crystal clear, you're building entirely my code (ie
> 3.1.0 + (something more than 1133) and not merging into an old master?
>
> Yes, I checked out your code as an entirely new branch, without merging
> anything into it. As far as I can tell, it's your code and your code alone.
>
> But just to make sure, I'll nuke my entire project directory and download
> a new copy. Something strange is obviously going on here, and I don't want
> to hold things up with incompetence ;-).
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/darktable-org/darktable/pull/4702#issuecomment-613285381>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AN2X2RWDUGHYK4CCI55IYQDRMQJILANCNFSM4MGRZLTQ>
> .
>


-- 
Graham Byrnes
Bron (Lyon), France
Mes photos: https://500px.com/grahambyrnes

GrahamByrnes commented:
I made a movie :-D

https://www.dropbox.com/s/s7qbe8vp73f8bsh/~_darktablePDtest_build%202020-04-14%2013-24-59.mp4?dl=0

bastibe commented:
I think the movie only shows your terminal, as far as I can tell?

GrahamByrnes commented:
Argh... speaking of incompetence lol
I'll try again with the whole screen then.

GrahamByrnes commented:
I was using the microsoft "game bar", which gives you a free screen recorder... but only allows a single window. It automatically shuts off when you exit a session, and it fails to show the preferences panel when I open it, you just see my cursor doing magic movements on the screen.

Anyway, the first video has preview_downsampling=0.25, I do stuff, reset it to 1.0, and close.
The second video shows what happens when I re-open at 1.0

https://www.dropbox.com/s/imck7li4qtegnrz/darktable%202020-04-14%2014-58-27.mp4?dl=0

https://www.dropbox.com/s/17i2vuyzdv80qwq/darktable%202020-04-14%2015-01-21.mp4?dl=0

MStraeten commented:
strange effects when using retouch using circle
![image](https://user-images.githubusercontent.com/39386816/79594490-5fb44980-80dd-11ea-855a-fb4d827f3251.png)
ellipse is fine

GrahamByrnes commented:
@MStraeten that's interesting... I see the same effect for each of circles, ellipses, paths... but all of them wrong, I think. In any case the cross is progressively further from the centre of the shape: I need to read the manual on retouch.
You're running a recent version ? Do you also have issues with the masks under exposure?
Thanks,
G

MStraeten commented:
Im running current master with just this pr added. The retouch function itself is fine - the correction is done at the mouse position not the displayed circe. After disabling circle and displaying the masks the circle i drawn at it’s correct place and can be adjusted there.
Same behaviour when using circle mask in exposure.

GrahamByrnes commented:
I'm finding different behaviours depending on the selected tool: I see strangeness if I place a shape without first selecting a tool (I'm looking at how to fix that), but not if I first select one of the tools.

I'm also running up against this story of shapes persisting in mask manager despite having deleted them...

jenshannoschwalm commented:
> I'm also running up against this story of shapes persisting in mask manager despite having deleted them..

This is about history compression. A mask created in a module is known to belong to that module via the history. If the history is compressed it is **very** tricky to keep the ownership (i worked on that a lot and could not find a safe solution as masks can be used by many modules ...) so it is tranferred to to the mask manager owning the mask. These masks can not be deleted due to the way the mask manager is inserted into the history at the very bottom.

GrahamByrnes commented:
So is this collection of masks bounded?

GrahamByrnes commented:
@MStraeten oh hang on, I see... your feather is _inside_ the circle... Hm.

MStraeten commented:
but thats only the displayed stuff - after it's applied the mask works fine (properly feathered) I'm not able to change the feather before applying a circle mask.

GrahamByrnes commented:
Yes... you have a 4K or 5K screen? OpenCL?

The only issue I have is the distance from the cross to the form. But seeing the feather somehow become negative on some machines is... strange.

Anything strange about gradients for you?

GrahamByrnes commented:
Are the the darkroom 2nd window settings used for 4K displays?

If so, I may have found the problem...

GrahamByrnes commented:
Just a short update:
a) I was being quite naive about how things work: in the case of the circle, the initial placement was in fact in a microscopic dot up in the top left corner. However once the mouse was clicked, all seemed well. After a bit of work to make the initial shape visible, it was obvious that there was something radical happening on clicking. Brush was similar, ellipse was fine, gradient was fine except that it was too wide.

b) The above things are all fixed... in somewhat strange ways. It seems that a backtransform is needed to write the mask onto the original data, but this re-scaled the radii of circles etc in the wrong way. So at the moment, they are back-transformed, then re-scaled explicitly with preview_downsampling *squared* to get the dimensions correct. 

c) Gradient was fixed with a couple of uses of preview_downsampling (not squared!). Ellipse continues to function blissfully without any intervention: I suspect because it uses a custom function to draw the ellipse, rather than going directly to cairo as do the other 3 masks.

d) I cleaned up some of the code in develop.c, which was taking advantage of logical accidents: it's now obvious that the rescaling only gets applied once in `dt_dev_get_zoom_scale()`. I've also moved the modifications to transform_plus & backtransform_plus inside the mutex, which seems to avoid some occasional weirdness.

e) The remaining issue is in retouch.c It all looks fine, but the distance from the original mask to the clone scales with 1/preview_downsampling. I actually prefer it bigger, but it should be the same across all settings. Since the same issue applies across all masks, I'll attempt to fix it inside masks.c

Hopefully some working code in a few days :-)

GrahamByrnes commented:
Sigh...

I think I'm going to admit defeat here.

First because when I thought I was getting closer, I just discovered that zooming breaks the alignment between the mouse path and the modified pixels... but more importantly, somehow the extra computational load of moving up and down from the reduced preview image more-or-less breaks zoom itself. When you zoom in, the image first shrinks to a much smaller zersion, then jumps back to the correctly zoomed size. This just makes zoom too tiring to use, but it's essential for fixing small dust-spots etc.

So, unless there are calls for clemency in the next couple of days, I'll close this. It's been an excellent way of learning how much I don't understand about the inner workings of the masks and the pixel-pipe!

chhil commented:
@GrahamByrnes 
Thank you for trying. I have seen your efforts and frustrations in my emails.

aurelienpierre commented:
Congrats on your persistence nonetheless.
